src/main/java/hudson/plugins/git/GitSCM.java

```diff
@@ -1325,14 +1332,38 @@ private void computeChangeLog(GitClient git, Revision revToBuild, TaskListener l
                 }
             }
 
+            if (context.getEnvironment().containsKey("CHANGE_TARGET")) {
+                listener.getLogger().println("Found CHANGE_TARGET in environment, considering it as origin.");
+                final String name = context.getEnvironment().get("CHANGE_TARGET");
+                git.fetch(getRepositories().get(0).getName(), new RefSpec("refs/heads/" + name + ":" + "refs/heads/" + name));
+                ObjectId rev = git.revParse(name);
+                changelog.excludes(rev);
+                exclusion = true;
+            }
+
             if (!exclusion) {
-                // this is the first time we are building this branch, so there's no base line to compare against.
-                // if we force the changelog, it'll contain all the changes in the repo, which is not what we want.
-                listener.getLogger().println("First time build. Skipping changelog.");
-            } else {
-                changelog.to(out).max(MAX_CHANGELOG).execute();
-                executed = true;
+                listener.getLogger().println("First build of this branch, computing origin.");
```

Comment:
Could you insert a comment that describes the algorithm being used to halt the search for the base of the current changelog?

I assume I'll decode the details of the algorithm while I'm doing the detailing code review and test scenario review, but it would be easier for me if there were a description of the technique inside the code.